### PR TITLE
Firewall DEH

### PR DESCRIPTION
**Pull request by jyvenugo**
- Parse the manifest for firewall rules and create the rule by calling the adding a firewall rule api.
- Tested by updating notepadplus.msix test package to include firewall rules and installing the same.
- Add and remove firewall rule in the format, PackageDisplayName (ProtocolName - Direction). so that each individual rule can be added and deleted uniquely.

### CODE REVIEW

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    return S_OK;
+}
+
+HRESULT FirewallRules::AddFirewallRules(FirewallRule& firewallRule)
+{
+    INetFwPolicy2 *pNetFwPolicy2 = NULL;
+    INetFwRules *pFwRules = NULL;
+    INetFwRule *pFwRule = NULL;
```
Comment (wcheng-msft) [inline preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: these should be wrapped in com pointers to release them when out of scope

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    // Create a new Firewall Rule object.
+    RETURN_IF_FAILED(CoCreateInstance(__uuidof(NetFwRule), NULL, CLSCTX_INPROC_SERVER, __uuidof(INetFwRule), (void**)&pFwRule));
+
+    std::wstring packageDisplayName = m_msixRequest->GetPackageInfo()->GetDisplayName();
+    BSTR ruleName = SysAllocString(packageDisplayName.data());
+    BSTR ruleDescription = SysAllocString(packageDisplayName.data());
+    pFwRule->put_Name(ruleName);
+    pFwRule->put_Description(ruleDescription);
+    pFwRule->put_Enabled(VARIANT_TRUE);
+
```
Comment (wcheng-msft) [inline preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: these return an HRESULT, wrap them in RETURN_IF_FAILED as well.

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            return mapping.value;
+        }
+    }
+    return NET_FW_PROFILE2_ALL;
+}
+
+HRESULT FirewallRules::WFCOMInitialize(INetFwPolicy2 ** ppNetFwPolicy2)
+{
+    RETURN_IF_FAILED(CoCreateInstance(__uuidof(NetFwPolicy2), NULL, CLSCTX_INPROC_SERVER, __uuidof(INetFwPolicy2), (void**)ppNetFwPolicy2));
+}
```
Comment (wcheng-msft) [inline preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: shouldn't this return S_OK;?

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    RETURN_IF_FAILED(localInstance->ParseManifest());
+
+    *instance = localInstance.release();
+
+    return S_OK;
+}
+
+HRESULT FirewallRules::ExecuteForRemoveRequest()
```
Comment (wcheng-msft) [inline preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: >ExecuteForRemoveRequest [](start = 23, length = 23)

implement this?

**Comment by wcheng-msft** [comment]
did you create tests / test packages for this?

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp (hunk 1 above)
Comment (jyvenugo) [reply preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: Updated to use com pointers.

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp (hunk 2 above)
Comment (jyvenugo) [reply preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: Used RETURN_IF_FAILED.

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp (hunk 3 above)
Comment (jyvenugo) [reply preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: Modified to return S_OK.

**Comment by jyvenugo** [comment]
I have not yet updated the powershell test script. Updated the package notepadplus.msix to include the following firewall rules and tested with the same,
    <desktop2:Extension Category="windows.firewallRules">
      <desktop2:FirewallRules Executable="VFS\ProgramFilesX86\Notepad++\notepad++.exe">
        <desktop2:Rule Direction="out" IPProtocol="GRE" Profile="domainAndPrivate" />
        <desktop2:Rule Direction="in" IPProtocol="TCP" Profile="private" LocalPortMin="2400" LocalPortMax="2450" />
        <desktop2:Rule Direction="in" IPProtocol="UDP" Profile="all" RemotePortMin="2400" RemotePortMax="2450" />
      </desktop2:FirewallRules>
    </desktop2:Extension>

File: preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp (hunk 4 above)
Comment (jyvenugo) [reply preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp]: Implemented ExecuteForRemoveRequest.

**Pull request merged**